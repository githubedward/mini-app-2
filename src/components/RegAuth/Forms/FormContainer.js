import React, { Component } from "react";
// component/styles
import LoginForm from "./LoginForm";
import SignupForm from "./SignupForm";
import styles from "./FormContainer.module.css";
// others
import * as helper from "../../../utils/functions";
import Api from "../../../services/index";

class FormContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      signupInput: {
        username: "",
        password: "",
        fullname: ""
      },
      loginInput: {
        username: "",
        password: ""
      },
      isFocus: {
        isSignupFocus: false,
        isLoginFocus: false
      },
      asyncSignupError: {},
      asyncLoginError: {}
    };
  }

  shouldComponentUpdate(nextProps, nextState) {
    return this.state !== nextState;
  }

  handleSignup = async data => {
    await helper.delay(500);
    try {
      // const resp = await axios.post(SIGNUP_URL, data);
      const resp = await Api.signup(data);
      const { username } = resp;
      this.setState({
        ...this.state,
        loginInput: {
          ...this.state.loginInput,
          username: username,
          password: ""
        },
        isFocus: {
          isSignupFocus: false,
          isLoginFocus: false
        }
      });
    } catch (err) {
      const error = err.response.data.errors[0];
      const asyncSignupError = {
        [error.path]: true,
        message: error.message
      };
      this.setState({
        ...this.state,
        asyncSignupError
      });
    }
  };

  handleLogin = async data => {
    await helper.delay(500);
    try {
      // const resp = await axios.post(LOGIN_URL, data);
      const resp = await Api.login(data);
      localStorage.setItem("token", resp.token);
      this.props.authenticateUserAction();
    } catch (err) {
      this.setState({
        ...this.state,
        asyncLoginError: err.response.data
      });
    }
  };

  // reset login inputs and disable login btn when signup input is onfocus
  handleSignupFocus = () => {
    if (!this.state.isFocus.isSignupFocus) {
      this.setState({
        ...this.state,
        isFocus: {
          isSignupFocus: true,
          isLoginFocus: false
        },
        loginInput: {
          username: "",
          password: ""
        },
        asyncLoginError: {}
      });
    }
  };

  // reset signup inputs and disable signup btn when login input is onfocus
  handleLoginFocus = () => {
    if (!this.state.isFocus.isLoginFocus) {
      this.setState({
        ...this.state,
        isFocus: {
          isSignupFocus: false,
          isLoginFocus: true
        },
        isSignupFocus: {
          username: "",
          password: "",
          fullname: ""
        },
        asyncSignupError: {}
      });
    }
  };

  render() {
    const {
      loginInput,
      signupInput,
      isFocus,
      asyncSignupError,
      asyncLoginError
    } = this.state;
    return (
      <section className={styles.background}>
        <div className={styles.logo}>MapSocial</div>
        <div className={styles.container}>
          <LoginForm
            onSubmit={this.handleLogin}
            user={loginInput}
            isFocus={isFocus}
            asyncError={asyncLoginError}
            handleLoginFocus={this.handleLoginFocus}
          />
          <SignupForm
            onSubmit={this.handleSignup}
            user={signupInput}
            isFocus={isFocus}
            asyncError={asyncSignupError}
            handleSignupFocus={this.handleSignupFocus}
          />
        </div>
      </section>
    );
  }
}

export default FormContainer;
