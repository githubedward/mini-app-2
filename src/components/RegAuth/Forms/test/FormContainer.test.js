import React from "react";
import FormContainer from "../FormContainer";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import Enzyme, { mount, unmount } from "enzyme";

Enzyme.configure({ adapter: new Adapter() });

describe("<FormContainer />", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<FormContainer />);
  });

  it("should render correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
