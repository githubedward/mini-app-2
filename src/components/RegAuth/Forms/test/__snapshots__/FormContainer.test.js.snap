// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormContainer /> should render correctly 1`] = `
<FormContainer>
  <section
    className="background"
  >
    <div
      className="logo"
    >
      MapSocial
    </div>
    <div
      className="container"
    >
      <WithFormik(LoginForm)
        asyncError={Object {}}
        handleLoginFocus={[Function]}
        isFocus={
          Object {
            "isLoginFocus": false,
            "isSignupFocus": false,
          }
        }
        onSubmit={[Function]}
        user={
          Object {
            "password": "",
            "username": "",
          }
        }
      >
        <Formik
          asyncError={Object {}}
          displayName="LoginForm"
          enableReinitialize={true}
          handleLoginFocus={[Function]}
          handleSubmit={[Function]}
          initialValues={
            Object {
              "asyncError": Object {},
              "handleLoginFocus": [Function],
              "isLoginFocus": false,
              "isSignupFocus": false,
              "password": "",
              "username": "",
            }
          }
          isFocus={
            Object {
              "isLoginFocus": false,
              "isSignupFocus": false,
            }
          }
          isInitialValid={false}
          onSubmit={[Function]}
          render={[Function]}
          user={
            Object {
              "password": "",
              "username": "",
            }
          }
          validateOnBlur={true}
          validateOnChange={true}
          validationSchema={[Function]}
        >
          <LoginForm
            asyncError={Object {}}
            dirty={false}
            errors={Object {}}
            handleBlur={[Function]}
            handleChange={[Function]}
            handleLoginFocus={[Function]}
            handleReset={[Function]}
            handleSubmit={[Function]}
            initialValues={
              Object {
                "asyncError": Object {},
                "handleLoginFocus": [Function],
                "isLoginFocus": false,
                "isSignupFocus": false,
                "password": "",
                "username": "",
              }
            }
            isFocus={
              Object {
                "isLoginFocus": false,
                "isSignupFocus": false,
              }
            }
            isSubmitting={false}
            isValid={false}
            isValidating={false}
            onSubmit={[Function]}
            registerField={[Function]}
            resetForm={[Function]}
            setError={[Function]}
            setErrors={[Function]}
            setFieldError={[Function]}
            setFieldTouched={[Function]}
            setFieldValue={[Function]}
            setFormikState={[Function]}
            setStatus={[Function]}
            setSubmitting={[Function]}
            setTouched={[Function]}
            setValues={[Function]}
            submitCount={0}
            submitForm={[Function]}
            touched={Object {}}
            unregisterField={[Function]}
            user={
              Object {
                "password": "",
                "username": "",
              }
            }
            validateField={[Function]}
            validateForm={[Function]}
            validateOnBlur={true}
            validateOnChange={true}
            values={
              Object {
                "asyncError": Object {},
                "handleLoginFocus": [Function],
                "isLoginFocus": false,
                "isSignupFocus": false,
                "password": "",
                "username": "",
              }
            }
          >
            <form
              className="form"
              onSubmit={[Function]}
            >
              <LoginInput
                label="Username"
                name="username"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                pressed={false}
                type="text"
                value=""
              >
                <label
                  className="label undefined false"
                >
                  Username
                  <input
                    className="input undefined false"
                    data-for="username-login"
                    name="username"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                    value=""
                  />
                </label>
              </LoginInput>
              <LoginInput
                label="Password"
                name="password"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                pressed={false}
                type="password"
                value=""
              >
                <label
                  className="label undefined false"
                >
                  Password
                  <input
                    className="input undefined false"
                    data-for="password-login"
                    name="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="password"
                    value=""
                  />
                </label>
              </LoginInput>
              <button
                className="btn_login"
                disabled={false}
                type="submit"
              >
                Login
              </button>
            </form>
          </LoginForm>
        </Formik>
      </WithFormik(LoginForm)>
      <WithFormik(SignupForm)
        asyncError={Object {}}
        handleSignupFocus={[Function]}
        isFocus={
          Object {
            "isLoginFocus": false,
            "isSignupFocus": false,
          }
        }
        onSubmit={[Function]}
        user={
          Object {
            "fullname": "",
            "password": "",
            "username": "",
          }
        }
      >
        <Formik
          asyncError={Object {}}
          displayName="SignupForm"
          enableReinitialize={true}
          handleSignupFocus={[Function]}
          handleSubmit={[Function]}
          initialValues={
            Object {
              "asyncError": Object {},
              "fullname": "",
              "handleSignupFocus": [Function],
              "isLoginFocus": false,
              "isSignupFocus": false,
              "password": "",
              "username": "",
            }
          }
          isFocus={
            Object {
              "isLoginFocus": false,
              "isSignupFocus": false,
            }
          }
          isInitialValid={false}
          onSubmit={[Function]}
          render={[Function]}
          user={
            Object {
              "fullname": "",
              "password": "",
              "username": "",
            }
          }
          validateOnBlur={true}
          validateOnChange={true}
          validationSchema={[Function]}
        >
          <SignupForm
            asyncError={Object {}}
            dirty={false}
            errors={Object {}}
            handleBlur={[Function]}
            handleChange={[Function]}
            handleReset={[Function]}
            handleSignupFocus={[Function]}
            handleSubmit={[Function]}
            initialValues={
              Object {
                "asyncError": Object {},
                "fullname": "",
                "handleSignupFocus": [Function],
                "isLoginFocus": false,
                "isSignupFocus": false,
                "password": "",
                "username": "",
              }
            }
            isFocus={
              Object {
                "isLoginFocus": false,
                "isSignupFocus": false,
              }
            }
            isSubmitting={false}
            isValid={false}
            isValidating={false}
            onSubmit={[Function]}
            registerField={[Function]}
            resetForm={[Function]}
            setError={[Function]}
            setErrors={[Function]}
            setFieldError={[Function]}
            setFieldTouched={[Function]}
            setFieldValue={[Function]}
            setFormikState={[Function]}
            setStatus={[Function]}
            setSubmitting={[Function]}
            setTouched={[Function]}
            setValues={[Function]}
            submitCount={0}
            submitForm={[Function]}
            touched={Object {}}
            unregisterField={[Function]}
            user={
              Object {
                "fullname": "",
                "password": "",
                "username": "",
              }
            }
            validateField={[Function]}
            validateForm={[Function]}
            validateOnBlur={true}
            validateOnChange={true}
            values={
              Object {
                "asyncError": Object {},
                "fullname": "",
                "handleSignupFocus": [Function],
                "isLoginFocus": false,
                "isSignupFocus": false,
                "password": "",
                "username": "",
              }
            }
          >
            <form
              className="form"
              onSubmit={[Function]}
            >
              <h1
                className="h1 false"
              >
                <strong>
                  Signup
                </strong>
                 to explore and share new experiences
              </h1>
              <SignupInput
                name="fullname"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Full Name"
                pressed={false}
                type="text"
                value=""
              >
                <label
                  className="label"
                >
                  <input
                    className="input undefined false"
                    data-for="fullname-signup"
                    name="fullname"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Full Name"
                    type="text"
                    value=""
                  />
                </label>
              </SignupInput>
              <SignupInput
                name="username"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Username"
                pressed={false}
                type="text"
                value=""
              >
                <label
                  className="label"
                >
                  <input
                    className="input undefined false"
                    data-for="username-signup"
                    name="username"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Username"
                    type="text"
                    value=""
                  />
                </label>
              </SignupInput>
              <SignupInput
                name="password"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Password"
                pressed={false}
                type="password"
                value=""
              >
                <label
                  className="label"
                >
                  <input
                    className="input undefined false"
                    data-for="password-signup"
                    name="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
                  />
                </label>
              </SignupInput>
              <button
                className="btn_signup"
                disabled={false}
                type="submit"
              >
                Signup
              </button>
            </form>
          </SignupForm>
        </Formik>
      </WithFormik(SignupForm)>
    </div>
  </section>
</FormContainer>
`;
